### PR TITLE
drivers: pwm: stm32: support counter modes 

### PR DESCRIPTION
**Pull request by GeorgeCGV**
Adds an ability to specify timer counter mode as DTS property of the timer.

PWM driver makes use of that to support center-aligned and up/down-counter mode.

### CODE REVIEW

File: drivers/pwm/pwm_stm32.c [hunk 1]
```diff
@@ -656,27 +676,36 @@ static void pwm_stm32_irq_config_func_##index(const struct device *dev)        \
...
-		.pclken = DT_INST_CLK(index, timer),                           \
-		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(index),		       \
-		CAPTURE_INIT(index)					       \
-	};                                                                     \
-									       \
-	DEVICE_DT_INST_DEFINE(index, &pwm_stm32_init, NULL,                    \
-			    &pwm_stm32_data_##index,                           \
-			    &pwm_stm32_config_##index, POST_KERNEL,            \
-			    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,                \
+#define DT_TIMER_COUNTER_MODE(inst)                                                  \
```
Comment (erwango) [inline drivers/pwm/pwm_stm32.c:679]: Optional: Since this property is common to stm32 timer subnodes (counter_timer could use it as well), This could also be provided in an `include/dt-bindings/timer/timer_stm32.h` (name to be confirmed as this is not a real zephyr timer driver)

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this addition, a couple of comments.

File: drivers/pwm/pwm_stm32.c [hunk 2]
```diff
@@ -660,27 +676,36 @@ static void pwm_stm32_irq_config_func_##index(const struct device *dev)        \
...
+	(DT_ENUM_IDX(inst, countermode) == 0) ? LL_TIM_COUNTERMODE_UP                \
+	: ((DT_ENUM_IDX(inst, countermode) == 1) ? LL_TIM_COUNTERMODE_DOWN           \
+	: ((DT_ENUM_IDX(inst, countermode) == 2) ? LL_TIM_COUNTERMODE_CENTER_DOWN    \
+	: ((DT_ENUM_IDX(inst, countermode) == 3) ? LL_TIM_COUNTERMODE_CENTER_UP      \
```
Comment (gmarull) [inline drivers/pwm/pwm_stm32.c:683]: Suggestion: you can also use an `int` field in DT and provide definitions via `include/dt-bindings/...`, e.g.

```
countermode = <STM32_TIM_COUNTERMODE_CENTER_UP>;
```

its value would be equal to `LL_...`.

**Comment by ABOSTM** [comment]
Hi @GeorgeCGV,
I don't catch the use case of counter mode support for `PWM` Zephyr API:
It would make sense to me to have management of counter mode in `counter` Zephyr API because application can access to counter value (`counter_get_value()`).
But for PWM I don't get the point ... maybe capture ?? Could you please detail ?

By the way setting a output PWM in downcounting mode is slightly different than for upcounting:
* upcounting: see RM0008 Rev 21: "The reference PWM signal OCxREF is high as long as TIMx_CNT < TIMx_CCRx"
* downcounting : "the reference signal OCxRef is low as long as TIMx_CNT > TIMx_CCRx"
 which is equivalent to OCxREF is high as long as TIMx_CNT =< TIMx_CCRx

The difference is when CNT is equal to CCRX.
In other words, in downcounting signal is high during 1 more tick than in upcounting, thus when PWM is configured in downcounting you need to change the formula to get the requested pulse duration.

Also in center aligned mode, I agree that you can divide the requested pulse length by 2 `pulse_cycles /= 2;` 
but this is not exact for the period:
as you can see in the picture below, total signal period is : [0 .. ARR] + [(ARR-1) .. 1]
The 1st part [0 .. ARR] is the one from upcounting, but the 2nd one is the one from upcounting minus 2.
Say it differently, in upcounting, `period = (ARR+1)`  but in center aligned, `period = 2 * ARR`
Thus computation of period should be adjusted when center aligned mode is used.

image

**Review by FRASTM — DISMISSED** [review]

File: drivers/pwm/pwm_stm32.c (hunk 2 above)
Comment (GeorgeCGV) [reply drivers/pwm/pwm_stm32.c:683]: It doesn't look like I can include `stm32_ll_tim.h` in the dt-bindings header.
But that is not a problem as the `CMS` value is similar between most SoCs.

**Comment by GeorgeCGV** [comment]
I stumbled upon the following case: the MP1XX reference manual contradicts LL definitions.

The issue is related to `LL_TIM_COUNTERMODE_CENTER_UP` and `LL_TIM_COUNTERMODE_CENTER_DOWN`.

Every `stm32ZYxx_ll_tim.h` defines them as ([link](https://github.com/STMicroelectronics/STM32CubeF4/blob/3d6be4bd406f275728e0a321cc371c62a3100533/Drivers/STM32F4xx_HAL_Driver/Inc/stm32f4xx_ll_tim.h#L567)):

- `LL_TIM_COUNTERMODE_CENTER_UP` `TIM_CR1_CMS1` - `0x40` (`b10`)
- `LL_TIM_COUNTERMODE_CENTER_DOWN` `TIM_CR1_CMS0` - `0x20` (`b01`)

And that matches with the reference manual.

But MP1XX defines them in a reversed way ([link](https://github.com/STMicroelectronics/STM32CubeMP1/blob/817e844fa791a34cf67df8100345f2a7382373d7/Drivers/STM32MP1xx_HAL_Driver/Inc/stm32mp1xx_ll_tim.h#L614))

- `LL_TIM_COUNTERMODE_CENTER_UP`  `TIM_CR1_CMS0` - `0x20` (`b01`) 
- `LL_TIM_COUNTERMODE_CENTER_DOWN` `TIM_CR1_CMS1` - `0x40` (`b10`)

That contradicts the reference manual of MP1XX.

I don't have a board with MP1XX to check. Is it a bug in the reference manual or LL defines are incorrect?

**Comment by GeorgeCGV** [comment]
Hello @ABOSTM, we use center-aligned PWM counter-mode for isolated DC/DC converter circuitry.

**Comment by ABOSTM** [comment]
@GeorgeCGV ,
You are right there is a misalignment between LL driver and Reference Manual of MP1. Most probably it is a LL bug.
I noticed that you submitted a PR on ST official repo. This is the right thing to do, but you need to sign the [Contribution License Agreement](https://cla.st.com/) otehrwise PR is automatically rejected.

> , we use a center-aligned timer counter-mode for isolated DC/DC converter circuitry that uses 2 PWM channels per converter.
 
OK, but it doesn't explain why center align mode is required. Can you elaborate ?

File: drivers/pwm/pwm_stm32.c (hunk 1 above)
Comment (GeorgeCGV) [reply drivers/pwm/pwm_stm32.c]: Switched to `int` constants that are defined in `include/dt-bindings/timer/stm32-timer.h`. Filename is composed based on the names of the other `dt-bindings` files (e.g. `pinctrl`).

**Comment by GeorgeCGV** [comment]
@ABOSTM, resubmitted the [PR](https://github.com/STMicroelectronics/STM32CubeMP1/pull/15) in [STM32CubeMP1](https://github.com/STMicroelectronics/STM32CubeMP1). Update this PR to remove the inverted defines of MP1XX.

> OK, but it doesn't explain why center align mode is required. Can you elaborate?

It was the easiest way to achieve the required PWM signals.

**Comment by ABOSTM** [comment]
> @ABOSTM, resubmitted the [PR](https://github.com/STMicroelectronics/STM32CubeMP1/pull/15) in [STM32CubeMP1](https://github.com/STMicroelectronics/STM32CubeMP1). Updated this PR to remove the inverted defines of MP1XX.
> 
> > OK, but it doesn't explain why center align mode is required. Can you elaborate?
> 
> It is the easiest way to achieve the required PWM signals. Why not use/hide what is provided by the hardware?

My point is: before this PR, it was already possible to generate PWM with Zephyr API (with underlying upcounting counter), so if we add some code (which is a cost in term of maintenance, Flash code size and runtime), it must be for some added value. 
At the end if you just need a PWM, I don't see the added value of center aligned mode.
In both cases (upcounting and center aligned), the PWM Zephyr API remain the same, so you will obtain the same PWM.

**Comment by GeorgeCGV** [comment]
> My point is: before this PR, it was already possible to generate PWM with Zephyr API (with underlying upcounting counter), so if we add some code (which is a cost in term of maintenance, Flash code size and runtime), it must be for some added value.
> At the end if you just need a PWM, I don't see the added value of center aligned mode.
> In both cases (upcounting and center aligned), the PWM Zephyr API remain the same, so you will obtain the same PWM.

It is confusing.  The Nordic PWM driver [provides](https://github.com/zephyrproject-rtos/zephyr/commit/4aaa08acfd53d9ed6002791a51cdb58e6e5c9425) a center-aligned mode for a couple of years.

Why do you say that the signals are the same for two different modes when the same API is used? 
There are differences between edge and center-aligned PWM signals.

**Comment by ABOSTM** [comment]
When using for example zephyr API  `pwm_pin_set_nsec(dev, pwm, period, pulse, flags)`, 
User expect to have the same waveform whatever the underlying implementation (upcounting, downcounting, or center aligned).
Now I can understand that interest is not really to countup or down or both, but rather to synchronise one PWM with another in the center of the pulse, Is it what you need ? If yes, then I am fine with this PR.
That said, there is still issue in downcounting about period computation, and similarly issue in center aligned about pulse computation.

**Comment by GeorgeCGV** [comment]
> synchronise one PWM with another in the center of the pulse, Is it what you need ?

@ABOSTM, yes. I am not sure what else can be expected in the case of timer mode regarding PWM. 
I didn't look into down-counting. The support of that mode can be added in another PR.

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: drivers/pwm/pwm_stm32.c [hunk 3]
```diff
@@ -238,6 +249,19 @@ static int pwm_stm32_pin_set(const struct device *dev, uint32_t pwm,
 		return -EINVAL;
 	}
 
+	if (cfg->countermode == LL_TIM_COUNTERMODE_UP) {
+		/* remove 1 cycle to account for 1 extra low cycle */
+		period_cycles -= 1U;
+	} else if (cfg->countermode == LL_TIM_COUNTERMODE_DOWN) {
+		/* add 1 low cycle to account for 1 extra high cycle */
+		period_cycles += 1U;
```
[the remark this answers is not in the document]
Comment (GeorgeCGV) [reply drivers/pwm/pwm_stm32.c:257]: @ABOSTM, shouldn't it be the case that down counting mode requires `-1` to the period (ARR) as well as `-1` to the pulse (CCR)? 

🟨 - getting rid of
🟩 - desired

**Upcounting**

> 0 up to ARR

_[0, ARR]_

`period_cycles = 8`
`pulse_cycles = 4`

| Level | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 |
|:-----:|---|---|---|---|---|---|---|---|---|
|   H   | 🟩 | 🟩 | 🟩 | 🟩 |   |   |   |   |   |
|   L   |   |   |   |   | 🟩 | 🟩 | 🟩 | 🟩 | 🟨 |

Pulse is `4` ticks, therefore CCR can be set to `pulse_cycles`. However, the period is reduced by `1` to be `[0, 7]` which is required `8` ticks.

`ARR = period_cycles - 1` (`ARR = 7`)
`CCR = pulse_cycles` (`CCR = 4`)

The end result is:

| Level | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 
|:-----:|---|---|---|---|---|---|---|---|
|   H   | 🟩 | 🟩 | 🟩 | 🟩 |   |   |   |   |
|   L   |   |   |   |   | 🟩 | 🟩 | 🟩 | 🟩 | 

**Downcounting**

> ARR down to 0

_[ARR, 0]_

`period_cycles = 8`
`pulse_cycles = 4`

| Level | 8 | 7 | 6 | 5 | 4 | 3 | 2 | 1 | 0 |
|:-----:|---|---|---|---|---|---|---|---|---|
|   H   |     |    |    |     |🟨 |🟩 |🟩|🟩 |🟩|
|   L    |🟩|🟩|🟩|🟩|    |     |    |    |     |

> the reference signal OCxRef is low as long as TIMx_CNT > TIMx_CCRx else it becomes high

To achieve desired `pulse_cycles` of `4` we set `CCR = pulse_cycles - 1` as you described ("So you must configure CCRx = pulse_cycles -1). 

That results in:

| Level | 8 | 7 | 6 | 5 | 4 | 3 | 2 | 1 | 0 |
|:-----:|---|---|---|---|---|---|---|---|---|
|   H   |     |    |    |     |    |🟩 |🟩|🟩 |🟩|
|   L    |🟩|🟩|🟩|🟩|🟩|     |    |    |     |

However, we also have to reduce the `ARR` by `1` as in the up-counting mode to get desired period of 8 ticks:

| Level | 7 | 6 | 5 | 4 | 3 | 2 | 1 | 0 |
|:-----:|---|---|---|---|---|---|---|---|
|   H   |    |    |     |    |🟩 |🟩|🟩 |🟩|
|   L    |🟩|🟩|🟩|🟩|     |    |    |     |

That yields:

`ARR = period_cycles - 1` (`ARR = 7`)
`CCR = pulse_cycles - 1` (`CCR = 3`)

Or?

[the remark this answers is not in the document]
Comment (ABOSTM) [reply drivers/pwm/pwm_stm32.c]: @GeorgeCGV 
My bad, you are right, I made a bad copy/paste:
Yes, In mean in down counting  `period_cycles -= 1U;`
And I fully agree with your new implementation

**Review by ABOSTM — DISMISSED** [review]
LGTM
Thanks

File: drivers/pwm/pwm_stm32.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (GeorgeCGV) [reply drivers/pwm/pwm_stm32.c]: @ABOSTM, thank you for the assistance

**Comment by slarti-pc** [comment]
Just for the use case:
Using two channels in center-aligned mode allows you some dead-time generation (often required for push-pull drivers to avoid excessive currents when switching over). While this approach is less flexible than the dedicated deadtime function in the advanced timers, the center-aligned mode is available also on basic timers, and might be sufficient.
"Normal" PWM mode can not be used for dead-time generation, since at the "period border" both channels will switch at exactly the same time.
And, of course, if you use/need only one output channel, center-aligned mode does not offer much benefit.

**Comment by ABOSTM** [comment]
@slarti-pc ,
ok that is clear, thanks for the explanation

**Review by ABOSTM — DISMISSED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]

File: dts/bindings/timer/st,stm32-timers.yaml [hunk 4]
```diff
@@ -26,3 +26,23 @@ properties:
...
+
+    st,countermode:
+        type: int
+        required: false
+        description: |
+          Sets timer counter mode.
+
+          Use constants defined in dt-bindings/timer/stm32-timer.h.
+
+          * STM32_TIM_COUNTERMODE_UP             - used as upcounter.
```
[the remark this answers is not in the document]
Comment (GeorgeCGV) [reply dts/bindings/timer/st,stm32-timers.yaml:39]: Specified reset state as the default value - STM32_TIM_COUNTERMODE_UP. Changed `DT_PROP_OR` to `DT_PROP` in the driver define.

**Review by erwango — DISMISSED** [review]

**Comment by mbolivar-nordic** [comment]
@erwango @ABOSTM could you please re-approve the rebase?

**Review by ABOSTM — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]
